{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "506bc897-ef7d-4d2d-abd6-5464adca6beb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/alexander/miniforge3/envs/tf_env/lib/python3.8/site-packages/openpyxl/worksheet/_read_only.py:79: UserWarning: Unknown extension is not supported and will be removed\n",
      "  for idx, row in parser.parse():\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "\n",
    "\n",
    "def get_codons_for_amino_acid(aa):\n",
    "    \"\"\"Return codons corresponding to a given amino acid.\"\"\"\n",
    "    # Dictionary mapping amino acids to their codons\n",
    "    codon_table = {\n",
    "        '*': ['TAA', 'TAG', 'TGA'],\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'M': ['ATG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC']\n",
    "    }\n",
    "\n",
    "    # Return codons for the provided amino acid\n",
    "    return codon_table.get(aa, None)\n",
    "\n",
    "\n",
    "def get_alternative_amino_acids(aa):\n",
    "    \"\"\"Return a list of alternative amino acids for the provided amino acid.\"\"\"\n",
    "    # List of all standard amino acids\n",
    "    standard_aas = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', '*']\n",
    "\n",
    "    # Check if the input amino acid is valid\n",
    "    if aa not in standard_aas:\n",
    "        raise ValueError(f\"'{aa}' is not a standard amino acid.\")\n",
    "\n",
    "    # Return all amino acids excluding the input one\n",
    "    return [a for a in standard_aas if a != aa]\n",
    "\n",
    "\n",
    "def is_point_mutation(current_codon, alt_codons):\n",
    "    \"\"\"\n",
    "    Check if codons in alt_codons differ from current_codon by a single point mutation.\n",
    "\n",
    "    Args:\n",
    "    - current_codon (str): The current codon.\n",
    "    - alt_codons (list): List of alternative codons.\n",
    "\n",
    "    Returns:\n",
    "    - List of codons from alt_codons that differ by a point mutation from current_codon.\n",
    "    \"\"\"\n",
    "    point_mutation_codons = []\n",
    "\n",
    "    for codon in alt_codons:\n",
    "        # Count differences between the current codon and the alternative codon\n",
    "        diff_count = sum(1 for a, b in zip(current_codon, codon) if a != b)\n",
    "\n",
    "        # If they differ by just one nucleotide, add to the result list\n",
    "        if diff_count == 1:\n",
    "            point_mutation_codons.append(codon)\n",
    "\n",
    "    return point_mutation_codons\n",
    "\n",
    "\n",
    "def is_tandem_double_mutation(current_codon, alt_codons):\n",
    "    \"\"\"\n",
    "    Check if any codons in alt_codons differ from current_codon by a tandem double nucleotide mutation.\n",
    "\n",
    "    Args:\n",
    "    - current_codon (str): The current codon.\n",
    "    - alt_codons (list): List of alternative codons.\n",
    "\n",
    "    Returns:\n",
    "    - List of codons from alt_codons that differ by a tandem double nucleotide mutation from current_codon.\n",
    "    \"\"\"\n",
    "    double_mutation_codons = []\n",
    "\n",
    "    for alt_codon in alt_codons:\n",
    "\n",
    "        # Identify positions where the codons differ\n",
    "        diff_positions = [i for i, (a, b) in enumerate(zip(current_codon, alt_codon)) if a != b]\n",
    "\n",
    "        # Check for tandem double mutations\n",
    "        if len(diff_positions) == 2 and diff_positions[1] - diff_positions[0] == 1:\n",
    "            double_mutation_codons.append(alt_codon)\n",
    "\n",
    "    return double_mutation_codons\n",
    "\n",
    "\n",
    "def calculate_fitness_average(SA_df, amber_pos, mutant_AA):\n",
    "    \"\"\"Calculate and return the average fitness.\"\"\"\n",
    "    filtered_df = SA_df[(SA_df[('Ambler Position', 'Unnamed: 1_level_1')] == amber_pos) & (SA_df[('Mutant AA', 'Unnamed: 5_level_1')] == mutant_AA)]\n",
    "    temp_list = filtered_df[('Fitness', 'Unnamed: 21_level_1')].tolist()\n",
    "    cleaned_list = [x for x in temp_list if not math.isnan(x)]\n",
    "    if len(cleaned_list) == 0:\n",
    "        return \"unknown\"\n",
    "    else:\n",
    "        return sum(cleaned_list)/len(cleaned_list)\n",
    "\n",
    "def main():\n",
    "    SA_df = pd.read_excel('../../Data_TEM1/14.xlsx', header=[0, 1])\n",
    "\n",
    "    WT_NT_seq_SA = [None] * (290 - 2)\n",
    "    WT_AA_seq_SA = [None] * (290 - 2)\n",
    "\n",
    "    for _, row in SA_df.iterrows():\n",
    "        amber_pos = row[('Ambler Position', 'Unnamed: 1_level_1')]\n",
    "        if amber_pos < 291:\n",
    "            WT_NT_seq_SA[amber_pos - 3] = row[('WT codon', 'Unnamed: 2_level_1')]\n",
    "            WT_AA_seq_SA[amber_pos - 3] = row[('WT AA', 'Unnamed: 4_level_1')]\n",
    "\n",
    "    mutation_data = {\n",
    "        'Allele': [], 'AA_wt': [], 'AA_variant': [], 'Position': [],\n",
    "        'fitness': [], 'mutation': [], 'codon': []\n",
    "    }\n",
    "\n",
    "    for c, (NT_t, AA_t) in enumerate(zip(WT_NT_seq_SA, WT_AA_seq_SA), start=3):\n",
    "        if AA_t:\n",
    "            for mutant_AA in get_alternative_amino_acids(AA_t):\n",
    "                mutation_data['Allele'].append(f\"{AA_t}{c}{mutant_AA}\")\n",
    "                mutation_data['AA_wt'].append(AA_t)\n",
    "                mutation_data['AA_variant'].append(mutant_AA)\n",
    "                mutation_data['Position'].append(c)\n",
    "                mutation_data['fitness'].append(calculate_fitness_average(SA_df, c, mutant_AA))\n",
    "                mutation_data['codon'].append(NT_t)\n",
    "\n",
    "                point_mutation_codons = is_point_mutation(NT_t, get_codons_for_amino_acid(mutant_AA))\n",
    "                tandem_mutation_codons = is_tandem_double_mutation(NT_t, get_codons_for_amino_acid(mutant_AA))\n",
    "                if point_mutation_codons:\n",
    "                    mutation_type = \"single\"\n",
    "                elif tandem_mutation_codons:\n",
    "                    mutation_type = \"double\"\n",
    "                else:\n",
    "                    mutation_type = \"NA\"\n",
    "\n",
    "                mutation_data['mutation'].append(mutation_type)\n",
    "\n",
    "    dfoutput = pd.DataFrame(mutation_data)\n",
    "    dfoutput.to_csv('../../Data_TEM1/data_TEM1_Codon_Substitutions.csv', index=False)\n",
    "\n",
    "\n",
    "def translate_codon(codon):\n",
    "    \"\"\"Translate a codon into its corresponding amino acid.\"\"\"\n",
    "    codon_table = {\n",
    "        '*': ['TAA', 'TAG', 'TGA'],\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'M': ['ATG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC']\n",
    "    }\n",
    "    for aa, codons in codon_table.items():\n",
    "        if codon in codons:\n",
    "            return aa\n",
    "    return None\n",
    "\n",
    "\n",
    "def get_alternative_amino_acids_combos(NT1, NT2):\n",
    "    \"\"\"Get amino acid combinations attainable by cross-codon tandem double mutations.\"\"\"\n",
    "    # Define the nucleotides\n",
    "    nucleotides = ['A', 'T', 'G', 'C']\n",
    "\n",
    "    # List to store amino acid combinations\n",
    "    aa_combinations = []\n",
    "\n",
    "    # Last nucleotide of NT1 and the first nucleotide of NT2\n",
    "    last_NT1, first_NT2 = NT1[-1], NT2[0]\n",
    "\n",
    "    # Original amino acids from the codons\n",
    "    original_aa1 = translate_codon(NT1)\n",
    "    original_aa2 = translate_codon(NT2)\n",
    "\n",
    "    # Iterate over possible mutations\n",
    "    for nuc1 in nucleotides:\n",
    "        for nuc2 in nucleotides:\n",
    "            if nuc1 != last_NT1 and nuc2 != first_NT2:\n",
    "                mutated_NT1 = NT1[:-1] + nuc1\n",
    "                mutated_NT2 = nuc2 + NT2[1:]\n",
    "\n",
    "                aa1 = translate_codon(mutated_NT1)\n",
    "                aa2 = translate_codon(mutated_NT2)\n",
    "\n",
    "                if aa1 and aa2 and (aa1 != original_aa1 and aa2 != original_aa2):\n",
    "                    aa_combinations.append((aa1, aa2))\n",
    "\n",
    "    unique_aa_combinations = list(set(aa_combinations))\n",
    "\n",
    "    return unique_aa_combinations\n",
    "\n",
    "\n",
    "def calculate_fitness_average_crosscodon(DA_df, amber_pos, mutant_AA1, mutant_AA2):\n",
    "    \"\"\"Calculate and return the average fitness.\"\"\"\n",
    "    filtered_df = DA_df[(DA_df[('Ambler Position')] == amber_pos) & (\n",
    "                DA_df[('Mut AA 1')] == mutant_AA1) & (\n",
    "                DA_df[('Mut AA 2')] == mutant_AA2)]\n",
    "\n",
    "    temp_list = filtered_df[('Double Mutant Fitness')].tolist()\n",
    "\n",
    "    if len(temp_list) == 0:\n",
    "        return \"unknown\"\n",
    "    else:\n",
    "        return sum(temp_list) / len(temp_list)\n",
    "\n",
    "\n",
    "def main2():\n",
    "    SA_df = pd.read_excel('../../Data_TEM1/14.xlsx', header=[0, 1])\n",
    "    DA_df = pd.read_excel(r'../../Data_TEM1/19.xlsx', sheet_name=1)\n",
    "\n",
    "    WT_NT_seq_SA = [None] * (290 - 2)\n",
    "    WT_AA_seq_SA = [None] * (290 - 2)\n",
    "\n",
    "    for _, row in SA_df.iterrows():\n",
    "        amber_pos = row[('Ambler Position', 'Unnamed: 1_level_1')]\n",
    "        if amber_pos < 291:\n",
    "            WT_NT_seq_SA[amber_pos - 3] = row[('WT codon', 'Unnamed: 2_level_1')]\n",
    "            WT_AA_seq_SA[amber_pos - 3] = row[('WT AA', 'Unnamed: 4_level_1')]\n",
    "\n",
    "    #print(WT_AA_seq_SA)\n",
    "    #print([index for index, value in enumerate(WT_AA_seq_SA) if value is None])\n",
    "\n",
    "    mutation_data = {\n",
    "        'Allele': [], 'AA_wt': [], 'AA_variant': [], 'Position': [],\n",
    "        'fitness': [], 'mutation': [], 'codon': []\n",
    "    }\n",
    "\n",
    "    codons = len(WT_AA_seq_SA)\n",
    "    for c in range(codons-1):\n",
    "        AA1_t = WT_AA_seq_SA[c]\n",
    "        AA2_t = WT_AA_seq_SA[c+1]\n",
    "\n",
    "        NT1_t = WT_NT_seq_SA[c]\n",
    "        NT2_t = WT_NT_seq_SA[c+1]\n",
    "\n",
    "        if NT1_t != None and NT2_t != None:\n",
    "            for mutant_AA_combo in get_alternative_amino_acids_combos(NT1_t, NT2_t):\n",
    "                mutant_AA1 = mutant_AA_combo[0]\n",
    "                mutant_AA2 = mutant_AA_combo[1]\n",
    "                mutation_data['Allele'].append(f\"{AA1_t}{AA2_t}{c+3}{mutant_AA1}{mutant_AA2}\")\n",
    "                mutation_data['AA_wt'].append(f\"{AA1_t}{AA2_t}\")\n",
    "                mutation_data['AA_variant'].append(f\"{mutant_AA1}{mutant_AA2}\")\n",
    "                mutation_data['Position'].append(c+3)\n",
    "                mutation_data['fitness'].append(calculate_fitness_average_crosscodon(DA_df, c+3, mutant_AA1, mutant_AA2))\n",
    "                mutation_data['codon'].append(f\"{NT1_t}{NT2_t}\")\n",
    "                mutation_data['mutation'].append(\"double\")\n",
    "\n",
    "    dfoutput = pd.DataFrame(mutation_data)\n",
    "    dfoutput.to_csv('../../Data_TEM1/data_TEM1_CrossCodonsDoubles.csv', index=False)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "32fa01f2-0537-4633-b8f4-d6bdbc43fb7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "len_1 2341\n",
      "len_1 50\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "'''Create singles_TEM1.npy for WF simulations'''\n",
    "\n",
    "file_path = '../../Data_TEM1/data_TEM1_Codon_Substitutions.csv'\n",
    "df = pd.read_csv(file_path)\n",
    "# Filter the rows\n",
    "filtered_fitness = df[(df['mutation'] == 'single') & (df['fitness'] != 'unknown')]['fitness']\n",
    "singles_fitness_arrays = np.array([float(item) if item != 'unknown' else np.nan for item in list(filtered_fitness)])\n",
    "np.save('../WFModel/WF_Simulations/singles_TEM1_new.npy', singles_fitness_arrays)\n",
    "\n",
    "\n",
    "'''Create doubles_TEM1.npy for WF simulations'''\n",
    "\n",
    "file_path1 = '../../Data_TEM1/data_TEM1_Codon_Substitutions.csv'\n",
    "file_path2 = '../../Data_TEM1/data_TEM1_CrossCodonsDoubles.csv'\n",
    "df1 = pd.read_csv(file_path1)\n",
    "# Filter the rows\n",
    "filtered_fitness1 = df1[(df1['mutation'] == 'double') & (df1['fitness'] != 'unknown')]['fitness']\n",
    "doubles_fitness_arrays1 = np.array([float(item) if item != 'unknown' else np.nan for item in list(filtered_fitness1)])\n",
    "print(\"len_1\", len(doubles_fitness_arrays1))\n",
    "df2 = pd.read_csv(file_path2)\n",
    "# Filter the rows\n",
    "filtered_fitness2 = df2[(df2['mutation'] == 'double') & (df2['fitness'] != 'unknown')]['fitness']\n",
    "doubles_fitness_arrays2 = np.array([float(item) if item != 'unknown' else np.nan for item in list(filtered_fitness2)])\n",
    "print(\"len_1\", len(doubles_fitness_arrays2))\n",
    "concatenated_doubles_fitness_arrays = np.concatenate((doubles_fitness_arrays1, doubles_fitness_arrays2))\n",
    "np.save('../WFModel/WF_Simulations/doubles_TEM1_new.npy', concatenated_doubles_fitness_arrays)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
