{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "899f21fa-9c2b-4cf5-9856-eddbfcf05a67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stats for TEM1\n",
      "\n",
      "Amino acids accessible through SN mutations 1758\n",
      "Amino acids accessible through DN mutations 4464\n",
      "Amino acids accessible only through DN mutations 2860\n",
      "One Amino acids accessible only through DN mutations 2460\n",
      "Two Amino acids accessible only through DN mutations 400\n",
      "\n",
      "Total distinct SN mutations 2574\n",
      "Expectation 2574\n",
      "\n",
      "Total distinct DN mutations 7713\n",
      "Expectation 7713\n",
      "\n",
      "--------------------------------------------------------\n",
      "--------------------------------------------------------\n",
      "\n",
      "Stats for P53\n",
      "\n",
      "Amino acids accessible through SN mutations 2437\n",
      "Amino acids accessible through DN mutations 6226\n",
      "Amino acids accessible only through DN mutations 3992\n",
      "One Amino acids accessible only through DN mutations 3419\n",
      "Two Amino acids accessible only through DN mutations 573\n",
      "\n",
      "Total distinct SN mutations 3546\n",
      "Expectation 3546\n",
      "\n",
      "Total distinct DN mutations 10629\n",
      "Expectation 10629\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "This script calculates the number of accessible amino acid sequences through SN and DN mutations, \n",
    "and those sequences that are exclusively accessible through DN mutations for both empirical landscapes.\n",
    "\"\"\"\n",
    "\n",
    "# Define DNA nucleotides as both a list and a set\n",
    "DNA_Nucleotides = ['A', 'C', 'G', 'T']\n",
    "DNA_Nucleotides_set = {'A', 'C', 'G', 'T'}\n",
    "\n",
    "# Dictionary for DNA reverse complement\n",
    "DNA_ReverseComplement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "\n",
    "# Dictionary mapping codon sequences to amino acids\n",
    "# '_' indicates a stop codon\n",
    "DNA_Codons = {\n",
    "    # 'M' - START, '_' - STOP\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
    "    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "    \"ATG\": \"M\",\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    \"TGG\": \"W\",\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
    "    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n",
    "}\n",
    "\n",
    "\n",
    "# Function to compute the number of different characters between two strings\n",
    "def differing_characters(s1, s2):\n",
    "    return sum(1 for a, b in zip(s1, s2) if a != b)\n",
    "\n",
    "\n",
    "# Function to convert a nucleotide sequence to an amino acid sequence\n",
    "def AASequence(ntSeq):\n",
    "    Codons = int(len(ntSeq)/3)\n",
    "    AA = []\n",
    "    for c in range(Codons):\n",
    "        codonSeq = ''.join(ntSeq[0+3*c:3+3*c])\n",
    "        AA.append(DNA_Codons[codonSeq])\n",
    "    return ''.join(AA)\n",
    "\n",
    "\n",
    "def print_stats(Wildtype):\n",
    "    \"\"\"\n",
    "    Analyze and print statistics on the number of accessible amino acid sequences\n",
    "    through single nucleotide (SN) and double nucleotide (DN) mutations.\n",
    "    \"\"\"\n",
    "    Codons = len(Wildtype)\n",
    "    Loci = Codons * 3\n",
    "    my_list = ''.join(Wildtype)\n",
    "    # Convert the string to a tuple of characters\n",
    "    wild_type = tuple(my_list)  # Original DNA sequence\n",
    "    AA = AASequence(wild_type)  # Amino acid sequence of the wild type\n",
    "\n",
    "    # Counters for various mutation types and their effects\n",
    "    counterSN = 0  # counter for AA accessibility through SN\n",
    "    counterDN = 0  # counter for AA accessibility through DN\n",
    "    counterDNMore = 0  # counter for AA accessibility through DN - SN\n",
    "    counterSNsyn = 0  # counter for synonymous SN mutation\n",
    "    counterDNsyn = 0  # counter for synonymous DN mutation\n",
    "\n",
    "    counterMutationsSN = 0\n",
    "\n",
    "    NewAASetSN = {AA}\n",
    "    # Iterate over each nucleotide position (locus) and consider single nucleotide mutations (SN)\n",
    "    for locus in range(Loci):  # iterating through all loci\n",
    "        substi = list(DNA_Nucleotides_set - {wild_type[locus]})  # all substitutions\n",
    "        copy_item = list(wild_type[:])\n",
    "        for s in range(len(substi)):  # iterating through all substitutions\n",
    "            counterMutationsSN += 1\n",
    "            copy_item[locus] = substi[s]\n",
    "            NewAA = AASequence(copy_item)  # mutant AA\n",
    "            # if locus >= 391:\n",
    "            #    print(NewAA[-1])\n",
    "            NewAASetSN.add(NewAA)\n",
    "            if AA == NewAA:  # Check for synonymous mutations\n",
    "                counterSNsyn += 1\n",
    "    counterSN = len(NewAASetSN) - 1  # Count unique amino acid sequences due to SN mutations\n",
    "\n",
    "    print(\"\\nAmino acids accessible through SN mutations\", counterSN)\n",
    "\n",
    "    counterMutationsDN = 0\n",
    "    counterMutationsDN_twoAAmutations = 0\n",
    "\n",
    "    # Generate all possible double nucleotide (DN) mutation combinations\n",
    "    dn_mutations = [[i, i + 1] for i in range(Codons * 3 - 1)]  # all DN combinations\n",
    "    NewAASetDN = {AA}  # wild type AA\n",
    "    NewTwoAASetDN = {AA}\n",
    "    NewOneAASetDN = {AA}\n",
    "\n",
    "    # Iterate over each DN mutation combination\n",
    "    for dn_comb in range(len(dn_mutations)):  # iterating through all DN combinations\n",
    "        substi0 = list(DNA_Nucleotides_set - {wild_type[dn_mutations[dn_comb][0]]})  # all substitutions\n",
    "        substi1 = list(DNA_Nucleotides_set - {wild_type[dn_mutations[dn_comb][1]]})  # all substitutions\n",
    "\n",
    "        copy_item = list(wild_type[:])\n",
    "        for s0 in range(len(substi0)):  # iterating through all substitutions\n",
    "            for s1 in range(len(substi1)):  # iterating through all substitutions\n",
    "                counterMutationsDN += 1\n",
    "                copy_item[dn_mutations[dn_comb][0]] = substi0[s0]\n",
    "                copy_item[dn_mutations[dn_comb][1]] = substi1[s1]\n",
    "                NewAA = AASequence(copy_item)  # mutant AA\n",
    "\n",
    "                diff_count = differing_characters(AA, NewAA)\n",
    "                if diff_count > 1:\n",
    "                    counterMutationsDN_twoAAmutations += 1\n",
    "                    NewTwoAASetDN.add(NewAA)\n",
    "                if diff_count == 1:\n",
    "                    NewOneAASetDN.add(NewAA)\n",
    "                NewAASetDN.add(NewAA)\n",
    "                if AA == NewAA:  # Check for synonymous mutations\n",
    "                    counterDNsyn += 1\n",
    "    counterDN = len(NewAASetDN) - 1  # Count unique amino acid sequences due to DN mutations\n",
    "    counterDNMore = len(NewAASetDN.difference(NewAASetSN))  # Amino acids accessible only by DN and not SN\n",
    "    TWOAAcounterDNMore = len(NewTwoAASetDN.difference(NewAASetSN))  # Amino acids accessible only by DN and not SN\n",
    "    OneAAcounterDNMore = len(NewOneAASetDN.difference(NewAASetSN))  # Amino acids accessible only by DN and not SN\n",
    "\n",
    "    print(\"Amino acids accessible through DN mutations\", counterDN)\n",
    "    print(\"Amino acids accessible only through DN mutations\", counterDNMore)\n",
    "\n",
    "    print(\"One Amino acids accessible only through DN mutations\", OneAAcounterDNMore)\n",
    "    print(\"Two Amino acids accessible only through DN mutations\", TWOAAcounterDNMore)\n",
    "\n",
    "    print(\"\\nTotal distinct SN mutations\", counterMutationsSN)\n",
    "    print(\"Expectation\", Codons * 3 * 3)\n",
    "\n",
    "    print(\"\\nTotal distinct DN mutations\", counterMutationsDN)\n",
    "    print(\"Expectation\", (Codons * 3 - 1) * 3 * 3)\n",
    "\n",
    "\n",
    "# DNA sequence data\n",
    "# TEM 1\n",
    "Wildtype_TEM1 = ['ATG', 'AGT', 'ATT', 'CAA', 'CAT', 'TTC', 'CGT', 'GTC', 'GCC', 'CTT', 'ATT', 'CCC', 'TTT', 'TTT', 'GCG', 'GCA', 'TTT', 'TGC', 'CTT', 'CCT', 'GTT', 'TTT', 'GCT', 'CAC', 'CCA', 'GAA', 'ACG', 'CTG', 'GTG', 'AAA', 'GTA', 'AAA', 'GAT', 'GCT', 'GAA', 'GAT', 'CAG', 'TTG', 'GGT', 'GCA', 'CGA', 'GTG', 'GGT', 'TAC', 'ATC', 'GAA', 'CTG', 'GAT', 'CTC', 'AAC', 'AGC', 'GGT', 'AAG', 'ATC', 'CTT', 'GAG', 'AGT', 'TTT', 'CGC', 'CCC', 'GAA', 'GAA', 'CGT', 'TTT', 'CCA', 'ATG', 'ATG', 'AGC', 'ACT', 'TTT', 'AAA', 'GTT', 'CTG', 'CTA', 'TGT', 'GGC', 'GCG', 'GTA', 'TTA', 'TCC', 'CGT', 'GTT', 'GAC', 'GCC', 'GGG', 'CAA', 'GAG', 'CAA', 'CTC', 'GGT', 'CGC', 'CGC', 'ATA', 'CAC', 'TAT', 'TCT', 'CAG', 'AAT', 'GAC', 'TTG', 'GTT', 'GAG', 'TAC', 'TCA', 'CCA', 'GTC', 'ACA', 'GAA', 'AAG', 'CAT', 'CTT', 'ACG', 'GAT', 'GGC', 'ATG', 'ACA', 'GTA', 'AGA', 'GAA', 'TTA', 'TGC', 'AGT', 'GCT', 'GCC', 'ATA', 'ACC', 'ATG', 'AGT', 'GAT', 'AAC', 'ACT', 'GCG', 'GCC', 'AAC', 'TTA', 'CTT', 'CTG', 'ACA', 'ACG', 'ATC', 'GGA', 'GGA', 'CCG', 'AAG', 'GAG', 'CTA', 'ACC', 'GCT', 'TTT', 'TTG', 'CAC', 'AAC', 'ATG', 'GGG', 'GAT', 'CAT', 'GTA', 'ACT', 'CGC', 'CTT', 'GAT', 'CGT', 'TGG', 'GAA', 'CCG', 'GAG', 'CTG', 'AAT', 'GAA', 'GCC', 'ATA', 'CCA', 'AAC', 'GAC', 'GAG', 'CGT', 'GAC', 'ACC', 'ACG', 'ATG', 'CCT', 'GCA', 'GCA', 'ATG', 'GCA', 'ACA', 'ACG', 'TTG', 'CGC', 'AAA', 'CTA', 'TTA', 'ACT', 'GGC', 'GAA', 'CTA', 'CTT', 'ACT', 'CTA', 'GCT', 'TCC', 'CGG', 'CAA', 'CAA', 'TTA', 'ATA', 'GAC', 'TGG', 'ATG', 'GAG', 'GCG', 'GAT', 'AAA', 'GTT', 'GCA', 'GGA', 'CCA', 'CTT', 'CTG', 'CGC', 'TCG', 'GCC', 'CTT', 'CCG', 'GCT', 'GGC', 'TGG', 'TTT', 'ATT', 'GCT', 'GAT', 'AAA', 'TCT', 'GGA', 'GCC', 'GGT', 'GAG', 'CGT', 'GGG', 'TCT', 'CGC', 'GGT', 'ATC', 'ATT', 'GCA', 'GCA', 'CTG', 'GGG', 'CCA', 'GAT', 'GGT', 'AAG', 'CCC', 'TCC', 'CGT', 'ATC', 'GTA', 'GTT', 'ATC', 'TAC', 'ACG', 'ACG', 'GGG', 'AGT', 'CAG', 'GCA', 'ACT', 'ATG', 'GAT', 'GAA', 'CGA', 'AAT', 'AGA', 'CAG', 'ATC', 'GCT', 'GAG', 'ATA', 'GGT', 'GCC', 'TCA', 'CTG', 'ATT', 'AAG', 'CAT', 'TGG']\n",
    "# P53\n",
    "Wildtype_P53 = ['ATG', 'GAG', 'GAG', 'CCG', 'CAG', 'TCA', 'GAT', 'CCT', 'AGC', 'GTC', 'GAG', 'CCC', 'CCT', 'CTG', 'AGT', 'CAG', 'GAA', 'ACA', 'TTT', 'TCA', 'GAC', 'CTA', 'TGG', 'AAA', 'CTA', 'CTT', 'CCT', 'GAA', 'AAC', 'AAC', 'GTT', 'CTG', 'TCC', 'CCC', 'TTG', 'CCG', 'TCC', 'CAA', 'GCA', 'ATG', 'GAT', 'GAT', 'TTG', 'ATG', 'CTG', 'TCC', 'CCG', 'GAC', 'GAT', 'ATT', 'GAA', 'CAA', 'TGG', 'TTC', 'ACT', 'GAA', 'GAC', 'CCA', 'GGT', 'CCA', 'GAT', 'GAA', 'GCT', 'CCC', 'AGA', 'ATG', 'CCA', 'GAG', 'GCT', 'GCT', 'CCC', 'CGC', 'GTG', 'GCC', 'CCT', 'GCA', 'CCA', 'GCA', 'GCT', 'CCT', 'ACA', 'CCG', 'GCG', 'GCC', 'CCT', 'GCA', 'CCA', 'GCC', 'CCC', 'TCC', 'TGG', 'CCC', 'CTG', 'TCA', 'TCT', 'TCT', 'GTC', 'CCT', 'TCC', 'CAG', 'AAA', 'ACC', 'TAC', 'CAG', 'GGC', 'AGC', 'TAC', 'GGT', 'TTC', 'CGT', 'CTG', 'GGC', 'TTC', 'TTG', 'CAT', 'TCT', 'GGG', 'ACA', 'GCC', 'AAG', 'TCT', 'GTG', 'ACT', 'TGC', 'ACG', 'TAC', 'TCC', 'CCT', 'GCC', 'CTC', 'AAC', 'AAG', 'ATG', 'TTT', 'TGC', 'CAA', 'CTG', 'GCC', 'AAG', 'ACC', 'TGC', 'CCT', 'GTG', 'CAG', 'CTG', 'TGG', 'GTT', 'GAT', 'TCC', 'ACA', 'CCC', 'CCG', 'CCC', 'GGC', 'ACC', 'CGC', 'GTC', 'CGC', 'GCC', 'ATG', 'GCC', 'ATC', 'TAC', 'AAG', 'CAG', 'TCA', 'CAG', 'CAC', 'ATG', 'ACG', 'GAG', 'GTT', 'GTG', 'AGG', 'CGC', 'TGC', 'CCC', 'CAC', 'CAT', 'GAG', 'CGC', 'TGC', 'TCA', 'GAT', 'AGC', 'GAT', 'GGT', 'CTG', 'GCC', 'CCT', 'CCT', 'CAG', 'CAT', 'CTT', 'ATC', 'CGA', 'GTG', 'GAA', 'GGA', 'AAT', 'TTG', 'CGT', 'GTG', 'GAG', 'TAT', 'TTG', 'GAT', 'GAC', 'AGA', 'AAC', 'ACT', 'TTT', 'CGA', 'CAT', 'AGT', 'GTG', 'GTG', 'GTG', 'CCC', 'TAT', 'GAG', 'CCG', 'CCT', 'GAG', 'GTT', 'GGC', 'TCT', 'GAC', 'TGT', 'ACC', 'ACC', 'ATC', 'CAC', 'TAC', 'AAC', 'TAC', 'ATG', 'TGT', 'AAC', 'AGT', 'TCC', 'TGC', 'ATG', 'GGC', 'GGC', 'ATG', 'AAC', 'CGG', 'AGG', 'CCC', 'ATC', 'CTC', 'ACC', 'ATC', 'ATC', 'ACA', 'CTG', 'GAA', 'GAC', 'TCC', 'AGT', 'GGT', 'AAT', 'CTA', 'CTG', 'GGA', 'CGG', 'AAC', 'AGC', 'TTT', 'GAG', 'GTG', 'CGT', 'GTT', 'TGT', 'GCC', 'TGT', 'CCT', 'GGG', 'AGA', 'GAC', 'CGG', 'CGC', 'ACA', 'GAG', 'GAA', 'GAG', 'AAT', 'CTC', 'CGC', 'AAG', 'AAA', 'GGG', 'GAG', 'CCT', 'CAC', 'CAC', 'GAG', 'CTG', 'CCC', 'CCA', 'GGG', 'AGC', 'ACT', 'AAG', 'CGA', 'GCA', 'CTG', 'CCC', 'AAC', 'AAC', 'ACC', 'AGC', 'TCC', 'TCT', 'CCC', 'CAG', 'CCA', 'AAG', 'AAG', 'AAA', 'CCA', 'CTG', 'GAT', 'GGA', 'GAA', 'TAT', 'TTC', 'ACC', 'CTT', 'CAG', 'ATC', 'CGT', 'GGG', 'CGT', 'GAG', 'CGC', 'TTC', 'GAG', 'ATG', 'TTC', 'CGA', 'GAG', 'CTG', 'AAT', 'GAG', 'GCC', 'TTG', 'GAA', 'CTC', 'AAG', 'GAT', 'GCC', 'CAG', 'GCT', 'GGG', 'AAG', 'GAG', 'CCA', 'GGG', 'GGG', 'AGC', 'AGG', 'GCT', 'CAC', 'TCC', 'AGC', 'CAC', 'CTG', 'AAG', 'TCC', 'AAA', 'AAG', 'GGT', 'CAG', 'TCT', 'ACC', 'TCC', 'CGC', 'CAT', 'AAA', 'AAA', 'CTC', 'ATG', 'TTC', 'AAG', 'ACA', 'GAA', 'GGG', 'CCT', 'GAC', 'TCA', 'GAC', 'TAG']\n",
    "\n",
    "print('Stats for TEM1')\n",
    "print_stats(Wildtype_TEM1)\n",
    "print('\\n--------------------------------------------------------')\n",
    "print('--------------------------------------------------------')\n",
    "print('\\nStats for P53')\n",
    "print_stats(Wildtype_P53)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b9c5bb6-ed86-468c-84c9-8709b7b0fc47",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
