{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6115fabf-77b5-4ca5-af29-027a6c0ef599",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting sense\n",
      "RingCodons 3\n",
      "\n",
      "Final Statistics:\n",
      "Total genotypes considered: 226981. Out of 262144\n",
      "Total single nucleotide mutations considered: 6128487. Out of 7077888\n",
      "Total double nucleotide mutations considered: 18385461. Out of 21233664\n",
      "\n",
      "P_{inac,non} = 0.5100136068134858\n",
      "P_{inac,mis} = 0.5034772156517633\n",
      "------------------------------------------\n",
      "\n",
      "Nonsynonymous sequences accessible by SN mutation per codon:  6.09836065573727\n",
      "Nonsynonymous sequences accessible by DN mutation per codon:  15.869671910864607\n",
      "Nonsynonymous sequences only accessible by DN mutation per codon:  10.231658156409809\n",
      "Nonsynonymous sequences only accessible by SN mutation per codon:  0.46034690128224615\n",
      "Nonsynonymous sequences accessible by SN and DN mutation per codon:  5.971347087788434\n",
      "------------------------------------------\n",
      "Missense sequences accessible by SN mutation per codon:  5.803278688524342\n",
      "Missense sequences accessible by DN mutation per codon:  14.973676210784909\n",
      "Missense sequences only accessible by DN mutation per codon:  9.629131953776472\n",
      "Missense sequences only accessible by SN mutation per codon:  0.45873443151605486\n",
      "Missense sequences accessible by SN and DN mutation per codon:  5.677877590341463\n",
      "------------------------------------------\n",
      "\n",
      "Fraction of non-sense mutatione (SN) 0.04189435336976321\n",
      "Fraction of non-sense mutatione (DN) 0.06008938258333582\n",
      "------------------------------------------\n",
      "Fraction of synonymous mutations (SN) 0.24408014571949\n",
      "Fraction of synonymous mutations (DN) 0.012461803378223696\n",
      "------------------------------------------\n",
      "Fraction of nonsynonymous mutations (SN) 0.75591985428051\n",
      "Fraction of nonsynonymous mutations (DN) 0.9875381966217763\n",
      "------------------------------------------\n",
      "Fraction of missense mutations (SN) 0.7140255009107468\n",
      "Fraction of missense mutations (DN) 0.9274488140384405\n",
      "Fraction of missense mutations (DN) only accessible by doubles 0.466949346551604\n",
      "------------------------------------------\n",
      "Fraction of 1AA mutations (DN) 0.8882518637960723\n",
      "Fraction of 2AA mutations (DN) 0.09928633282570397\n",
      "Nonsynonymous sequences (1AA) only accessible by DN mutation per codon:  8.78688524590151\n",
      "Nonsynonymous sequences (2AA) only accessible by DN mutation per codon:  1.4447729105075076\n",
      "------------------------------------------\n",
      "Counter missense mutations (SN) 4375896\n",
      "Counter nonsense mutations (SN) 256749\n",
      "Counter synonymous mutations (SN) 1495842\n",
      "------------------------------------------\n",
      "Counter missense mutations (DN) 17051574\n",
      "Counter nonsense mutations (DN) 1104771\n",
      "Counter synonymous mutations (DN) 229116\n",
      "Counter missense mutations (DN) only accessible by DN 8585079\n",
      "Counter nonsense mutations (DN) only accessible by DN 674904\n",
      "------------------------------------------\n",
      "Sanity Check\n",
      "counter_all_SN: 6128487, counterSNsyn+counterSNmis+counterSNnon 6128487\n",
      "counter_all_DN: 18385461, counterDNsyn+counterDNmis+counterDNnon 18385461\n"
     ]
    }
   ],
   "source": [
    "from itertools import product\n",
    "import timeit\n",
    "\n",
    "print(\"Starting sense\")\n",
    "\n",
    "# Lists of DNA nucleotides and their respective sets.\n",
    "DNA_Nucleotides = ['A', 'C', 'G', 'T']\n",
    "DNA_Nucleotides_set = {'A', 'C', 'G', 'T'}\n",
    "\n",
    "# Dictionary for DNA nucleotide complement mapping.\n",
    "DNA_ReverseComplement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}\n",
    "\n",
    "# Dictionary of DNA codons to corresponding amino acids.\n",
    "DNA_Codons = {\n",
    "    # 'M' - START, '_' - STOP\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
    "    \"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "    \"ATG\": \"M\",\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "    \"TGG\": \"W\",\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
    "    \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"\n",
    "}\n",
    "\n",
    "\n",
    "def AASequence(ntSeq):\n",
    "    \"\"\"Convert a nucleotide sequence to its corresponding amino acid sequence.\"\"\"\n",
    "    Codons = int(len(ntSeq)/3)\n",
    "    AA = []\n",
    "    for c in range(Codons):\n",
    "        codonSeq = ''.join(ntSeq[0+3*c:3+3*c])\n",
    "        AA.append(DNA_Codons[codonSeq])\n",
    "    return ''.join(AA)\n",
    "\n",
    "\n",
    "def count_differences(seq1, seq2):\n",
    "    \"\"\"Count the number of differences between two sequences of the same length.\"\"\"\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(\"Sequences must be of the same length\")\n",
    "    return sum(1 for a, b in zip(seq1, seq2) if a != b)\n",
    "\n",
    "\n",
    "# Setup for analysis.\n",
    "Codons = 3\n",
    "print(\"RingCodons\", Codons)\n",
    "\n",
    "# Generate all possible DN combinations for the given number of codons.\n",
    "dn_mutations = [[i,i+1] for i in range(Codons*3-1)]\n",
    "dn_mutations.append([0, Codons*3-1])\n",
    "\n",
    "# Initializing counters to track various mutation outcomes.\n",
    "counterSN = 0\n",
    "counterSNOnlyMissense = 0\n",
    "counterDN = 0\n",
    "counterDNMore = 0\n",
    "counterSNMore = 0\n",
    "counterSNandDN = 0\n",
    "counterSNsyn = 0\n",
    "counterDNsyn = 0\n",
    "counterSNmis = 0\n",
    "counterDNmis = 0\n",
    "counterSNnon = 0\n",
    "counterDNnon = 0\n",
    "counter_all = 0\n",
    "counter_all_SN = 0\n",
    "counter_all_DN = 0\n",
    "counterOnlyDNacc = 0\n",
    "counterDNOnlyMissense = 0\n",
    "counterDNMoreOnlyMissense = 0\n",
    "counterSNMoreOnlyMissense = 0\n",
    "counterSNandDNOnlyMissense = 0\n",
    "counterOnlyDNaccMissense = 0\n",
    "counterDNmis1AA = 0\n",
    "counterDNnon1AA = 0\n",
    "counterDNmis2AA = 0\n",
    "counterDNnon2AA = 0\n",
    "counterDNMore1AA = 0\n",
    "counterDNMore2AA = 0\n",
    "\n",
    "\n",
    "# Main loop to iterate through all possible DNA sequences.\n",
    "start = timeit.default_timer()\n",
    "ci = 0\n",
    "for item in product(DNA_Nucleotides, repeat=3*Codons):\n",
    "    ci += 1\n",
    "    if ci % 1_000_000 == 0:\n",
    "        print(f\"Progress: {ci}, Time: {timeit.default_timer() - start}\")\n",
    "\n",
    "    # Process single nucleotide (SN) mutations.\n",
    "    AA = AASequence(item)  # Initial amino acid sequence.\n",
    "    if AA.find('_') == -1:  # Ignore initial sequences with stop codons.\n",
    "        counter_all += 1\n",
    "        NewAASetSN = {AA}\n",
    "        NewAASetSNOnlyMissense = {AA}\n",
    "        for locus in range(len(item)):  # iterating through all loci\n",
    "            substi = list(DNA_Nucleotides_set-{item[locus]})  # create all substitutions\n",
    "            copy_item = list(item[:])\n",
    "            for s in range(len(substi)):  # iterating through all substitutions\n",
    "                counter_all_SN += 1\n",
    "                copy_item[locus] = substi[s]\n",
    "                NewAA = AASequence(copy_item)  # mutant AA\n",
    "                NewAASetSN.add(NewAA)\n",
    "                # Classify the mutation based on its effect.\n",
    "                if AA == NewAA:\n",
    "                    counterSNsyn += 1\n",
    "                elif NewAA.find('_') == -1:\n",
    "                    counterSNmis += 1\n",
    "                    NewAASetSNOnlyMissense.add(NewAA)\n",
    "                else:\n",
    "                    counterSNnon += 1\n",
    "        counterSN += (len(NewAASetSN)-1)/Codons  # how many distinct mutant AA are accessible?\n",
    "        counterSNOnlyMissense += (len(NewAASetSNOnlyMissense)-1)/Codons  # how many distinct mutant AA are accessible?\n",
    "\n",
    "        # Process double nucleotide (DN) mutations.\n",
    "        NewAASetDN = {AA}\n",
    "        NewAASetDNOnlyMissense = {AA}\n",
    "        NewAASetDN1AA = {AA}\n",
    "        NewAASetDN2AA = {AA}\n",
    "\n",
    "        for dn_comb in range(len(dn_mutations)):  # iterating through all DN combinations\n",
    "            substi0 = list(DNA_Nucleotides_set-{item[dn_mutations[dn_comb][0]]})  # create all substitutions\n",
    "            substi1 = list(DNA_Nucleotides_set-{item[dn_mutations[dn_comb][1]]})  # create all substitutions\n",
    "            copy_item = list(item).copy()\n",
    "            for s0 in range(len(substi0)):  # iterating through all substitutions\n",
    "                for s1 in range(len(substi1)):  # iterating through all substitutions\n",
    "                    counter_all_DN += 1\n",
    "                    copy_item[dn_mutations[dn_comb][0]] = substi0[s0]\n",
    "                    copy_item[dn_mutations[dn_comb][1]] = substi1[s1]\n",
    "                    NewAA = AASequence(copy_item)  # mutant AA\n",
    "                    NewAASetDN.add(NewAA)\n",
    "                    if count_differences(NewAA, AA) == 1:\n",
    "                        NewAASetDN1AA.add(NewAA)\n",
    "                    elif count_differences(NewAA, AA) == 2:\n",
    "                        NewAASetDN2AA.add(NewAA)\n",
    "                    # Classify the mutation based on its effect.\n",
    "                    if AA == NewAA:\n",
    "                        counterDNsyn += 1\n",
    "                    elif NewAA.find('_') == -1:\n",
    "                        NewAASetDNOnlyMissense.add(NewAA)\n",
    "                        counterDNmis += 1\n",
    "                        if count_differences(NewAA, AA) == 1:\n",
    "                            counterDNmis1AA += 1\n",
    "                        elif count_differences(NewAA, AA) == 2:\n",
    "                            counterDNmis2AA += 1\n",
    "                        if NewAA not in NewAASetSN:\n",
    "                            counterOnlyDNaccMissense += 1\n",
    "                    else:\n",
    "                        counterDNnon += 1\n",
    "                        if count_differences(NewAA, AA) == 1:\n",
    "                            counterDNnon1AA += 1\n",
    "                        elif count_differences(NewAA, AA) == 2:\n",
    "                            counterDNnon2AA += 1\n",
    "                    if NewAA not in NewAASetSN:\n",
    "                        counterOnlyDNacc += 1\n",
    "        counterDN += (len(NewAASetDN)-1)/Codons\n",
    "        counterDNMore += (len(NewAASetDN.difference(NewAASetSN)))/Codons\n",
    "        counterSNMore += (len(NewAASetSN.difference(NewAASetDN)))/Codons\n",
    "        counterSNandDN += (len(NewAASetSN.intersection(NewAASetDN)))/Codons\n",
    "        counterDNOnlyMissense += (len(NewAASetDNOnlyMissense)-1)/Codons\n",
    "        counterDNMoreOnlyMissense += (len(NewAASetDNOnlyMissense.difference(NewAASetSNOnlyMissense)))/Codons\n",
    "        counterSNMoreOnlyMissense += (len(NewAASetSNOnlyMissense.difference(NewAASetDNOnlyMissense)))/Codons\n",
    "        counterSNandDNOnlyMissense += (len(NewAASetSNOnlyMissense.intersection(NewAASetDNOnlyMissense)))/Codons\n",
    "        counterDNMore1AA += (len(NewAASetDN1AA.difference(NewAASetSN)))/Codons\n",
    "        counterDNMore2AA += (len(NewAASetDN2AA.difference(NewAASetSN)))/Codons\n",
    "\n",
    "\n",
    "print(\"\\nFinal Statistics:\")\n",
    "print('Total genotypes considered: {}. Out of {}'.format(counter_all, 4**(Codons*3)))\n",
    "print('Total single nucleotide mutations considered: {}. Out of {}'.format(counter_all_SN, 4**(Codons*3)*Codons*3*3))\n",
    "print('Total double nucleotide mutations considered: {}. Out of {}'.format(counter_all_DN, 4**(Codons*3)*(Codons*3)*3*3))\n",
    "\n",
    "print('\\nP_{inac,non} =', counterOnlyDNacc/(counterDNmis+counterDNnon))\n",
    "print('P_{inac,mis} =', counterOnlyDNaccMissense/(counterDNmis))\n",
    "\n",
    "print('------------------------------------------')\n",
    "print(\"\\nNonsynonymous sequences accessible by SN mutation per codon: \", counterSN/counter_all)\n",
    "print(\"Nonsynonymous sequences accessible by DN mutation per codon: \", counterDN/counter_all)\n",
    "print(\"Nonsynonymous sequences only accessible by DN mutation per codon: \", counterDNMore/counter_all)\n",
    "print(\"Nonsynonymous sequences only accessible by SN mutation per codon: \", counterSNMore/counter_all)\n",
    "print(\"Nonsynonymous sequences accessible by SN and DN mutation per codon: \", counterSNandDN/counter_all)\n",
    "print('------------------------------------------')\n",
    "print(\"Missense sequences accessible by SN mutation per codon: \", counterSNOnlyMissense/counter_all)\n",
    "print(\"Missense sequences accessible by DN mutation per codon: \", counterDNOnlyMissense/counter_all)\n",
    "print(\"Missense sequences only accessible by DN mutation per codon: \", counterDNMoreOnlyMissense/counter_all)\n",
    "print(\"Missense sequences only accessible by SN mutation per codon: \", counterSNMoreOnlyMissense/counter_all)\n",
    "print(\"Missense sequences accessible by SN and DN mutation per codon: \", counterSNandDNOnlyMissense/counter_all)\n",
    "print('------------------------------------------')\n",
    "\n",
    "print(\"\\nFraction of non-sense mutatione (SN)\", counterSNnon/(counterSNsyn+counterSNmis+counterSNnon))\n",
    "print(\"Fraction of non-sense mutatione (DN)\", counterDNnon/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print('------------------------------------------')\n",
    "print(\"Fraction of synonymous mutations (SN)\", counterSNsyn/counter_all_SN)\n",
    "print(\"Fraction of synonymous mutations (DN)\", counterDNsyn/counter_all_DN)\n",
    "print('------------------------------------------')\n",
    "print(\"Fraction of nonsynonymous mutations (SN)\", (counterSNmis+counterSNnon)/(counterSNsyn+counterSNmis+counterSNnon))\n",
    "print(\"Fraction of nonsynonymous mutations (DN)\", (counterDNmis+counterDNnon)/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print('------------------------------------------')\n",
    "print(\"Fraction of missense mutations (SN)\", (counterSNmis)/(counterSNsyn+counterSNmis+counterSNnon))\n",
    "print(\"Fraction of missense mutations (DN)\", (counterDNmis)/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print(\"Fraction of missense mutations (DN) only accessible by doubles\", (counterOnlyDNaccMissense)/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print('------------------------------------------')\n",
    "print(\"Fraction of 1AA mutations (DN)\", (counterDNmis1AA+counterDNnon1AA)/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print(\"Fraction of 2AA mutations (DN)\", (counterDNmis2AA+counterDNnon2AA)/(counterDNsyn+counterDNmis+counterDNnon))\n",
    "print(\"Nonsynonymous sequences (1AA) only accessible by DN mutation per codon: \", counterDNMore1AA/counter_all)\n",
    "print(\"Nonsynonymous sequences (2AA) only accessible by DN mutation per codon: \", counterDNMore2AA/counter_all)\n",
    "print('------------------------------------------')\n",
    "print('Counter missense mutations (SN)', counterSNmis)\n",
    "print('Counter nonsense mutations (SN)', counterSNnon)\n",
    "print('Counter synonymous mutations (SN)', counterSNsyn)\n",
    "print('------------------------------------------')\n",
    "print('Counter missense mutations (DN)', counterDNmis)\n",
    "print('Counter nonsense mutations (DN)', counterDNnon)\n",
    "print('Counter synonymous mutations (DN)', counterDNsyn)\n",
    "print('Counter missense mutations (DN) only accessible by DN', counterOnlyDNaccMissense)\n",
    "print('Counter nonsense mutations (DN) only accessible by DN', counterOnlyDNacc-counterOnlyDNaccMissense)\n",
    "print('------------------------------------------')\n",
    "print('Sanity Check')\n",
    "print('counter_all_SN: {}, counterSNsyn+counterSNmis+counterSNnon {}'.format(counter_all_SN, counterSNsyn+counterSNmis+counterSNnon))\n",
    "print('counter_all_DN: {}, counterDNsyn+counterDNmis+counterDNnon {}'.format(counter_all_DN, counterDNsyn+counterDNmis+counterDNnon))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f4daeb-0ef7-47d0-9984-af014d21c73f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
